In [64]:
import math
import warnings

%matplotlib inline
import numpy as np
import pandas as pd
import scipy
import sklearn
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn import linear_model
import statsmodels.formula.api as smf

# Display preferences.
pd.options.display.float_format = '{:.3f}'.format

# Suppress annoying harmless error.
warnings.filterwarnings(
    action="ignore",
    module="scipy",
    message="^internal gelsd"
)

In [65]:
df = pd.read_excel("/Users/saurabh/Documents/PythonData/table_8_offenses_known_to_law_enforcement_california_by_city_2013.xls", encoding = "ISO-8859-1", skiprows=4)

In [66]:
df.head()

City  Population  Violent\ncrime  \
0      Adelanto   31165.000         198.000   
1  Agoura Hills   20762.000          19.000   
2       Alameda   76206.000         158.000   
3        Albany   19104.000          29.000   
4      Alhambra   84710.000         163.000   

   Murder and\nnonnegligent\nmanslaughter  Rape\n(revised\ndefinition)1  \
0                                   2.000                           nan   
1                                   0.000                           nan   
2                                   0.000                           nan   
3                                   0.000                           nan   
4                                   1.000                           nan   

   Rape\n(legacy\ndefinition)2  Robbery  Aggravated\nassault  Property\ncrime  \
0                       15.000   52.000              129.000          886.000   
1                        2.000   10.000                7.000          306.000   
2                       10.000   85.000               63.000         1902.000   
3                        1.000   24.000                4.000          557.000   
4                        9.000   81.000               72.000         1774.000   

   Burglary  Larceny-\ntheft  Motor\nvehicle\ntheft  Arson  
0   381.000          372.000                133.000 17.000  
1   109.000          185.000                 12.000  7.000  
2   287.000         1285.000                330.000 17.000  
3    94.000          388.000                 75.000  7.000  
4   344.000         1196.000                234.000  7.000

In [67]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 464 entries, 0 to 463
Data columns (total 13 columns):
City                                    464 non-null object
Population                              462 non-null float64
Violent
crime                           462 non-null float64
Murder and
nonnegligent
manslaughter    462 non-null float64
Rape
(revised
definition)1              0 non-null float64
Rape
(legacy
definition)2               462 non-null float64
Robbery                                 462 non-null float64
Aggravated
assault                      462 non-null float64
Property
crime                          462 non-null float64
Burglary                                462 non-null float64
Larceny-
theft                          462 non-null float64
Motor
vehicle
theft                     462 non-null float64
Arson                                   462 non-null float64
dtypes: float64(12), object(1)
memory usage: 47.2+ KB


In [68]:
df.columns

Index(['City', 'Population', 'Violent\ncrime',
       'Murder and\nnonnegligent\nmanslaughter',
       'Rape\n(revised\ndefinition)1', 'Rape\n(legacy\ndefinition)2',
       'Robbery', 'Aggravated\nassault', 'Property\ncrime', 'Burglary',
       'Larceny-\ntheft', 'Motor\nvehicle\ntheft', 'Arson'],
      dtype='object')

In [69]:
df_temp = df.rename(index=str, columns={"Property\ncrime": "Property_crime"})

In [70]:
df_temp.head()

City  Population  Violent\ncrime  \
0      Adelanto   31165.000         198.000   
1  Agoura Hills   20762.000          19.000   
2       Alameda   76206.000         158.000   
3        Albany   19104.000          29.000   
4      Alhambra   84710.000         163.000   

   Murder and\nnonnegligent\nmanslaughter  Rape\n(revised\ndefinition)1  \
0                                   2.000                           nan   
1                                   0.000                           nan   
2                                   0.000                           nan   
3                                   0.000                           nan   
4                                   1.000                           nan   

   Rape\n(legacy\ndefinition)2  Robbery  Aggravated\nassault  Property_crime  \
0                       15.000   52.000              129.000         886.000   
1                        2.000   10.000                7.000         306.000   
2                       10.000   85.000               63.000        1902.000   
3                        1.000   24.000                4.000         557.000   
4                        9.000   81.000               72.000        1774.000   

   Burglary  Larceny-\ntheft  Motor\nvehicle\ntheft  Arson  
0   381.000          372.000                133.000 17.000  
1   109.000          185.000                 12.000  7.000  
2   287.000         1285.000                330.000 17.000  
3    94.000          388.000                 75.000  7.000  
4   344.000         1196.000                234.000  7.000

In [75]:
df_temp.columns

Index(['City', 'Population', 'Violent\ncrime',
       'Murder and\nnonnegligent\nmanslaughter',
       'Rape\n(revised\ndefinition)1', 'Rape\n(legacy\ndefinition)2',
       'Robbery', 'Aggravated\nassault', 'Property_crime', 'Burglary',
       'Larceny-\ntheft', 'Motor\nvehicle\ntheft', 'Arson',
       'population_square', 'Robbery_new', 'Burglary_new', 'murder_new'],
      dtype='object')

In [76]:
df_temp.fillna(df_temp.mean(),inplace=True)

In [77]:
df_temp.info()

<class 'pandas.core.frame.DataFrame'>
Index: 464 entries, 0 to 463
Data columns (total 17 columns):
City                                    464 non-null object
Population                              464 non-null float64
Violent
crime                           464 non-null float64
Murder and
nonnegligent
manslaughter    464 non-null float64
Rape
(revised
definition)1              0 non-null float64
Rape
(legacy
definition)2               464 non-null float64
Robbery                                 464 non-null float64
Aggravated
assault                      464 non-null float64
Property_crime                          464 non-null float64
Burglary                                464 non-null float64
Larceny-
theft                          464 non-null float64
Motor
vehicle
theft                     464 non-null float64
Arson                                   464 non-null float64
population_square                       0 non-null float64
Robbery_new                             464 non-nul

In [78]:
df_temp['population_square'] = df_temp['Population'] * df_temp['Population']
df_temp['Robbery_new'] = np.where(df_temp['Robbery']>0, 1, 0)
df_temp['Burglary_new'] = np.where(df_temp['Burglary']>0, 1, 0)
df_temp['murder_new'] = np.where(df_temp.iloc[:,3]>0, 1,0)

In [79]:
# Instantiate and fit our model.
regr = linear_model.LinearRegression()
Y = df_temp['Property_crime'].values.reshape(-1, 1)
X = df_temp[['Population','population_square','Robbery_new','Burglary_new','murder_new']]
regr.fit(X, Y)

# Inspect the results.
print('\nCoefficients: \n', regr.coef_)
print('\nIntercept: \n', regr.intercept_)
print('\nR-squared:')
print(regr.score(X, Y))


Coefficients: 
 [[ 3.52427802e-02 -3.37625950e-09 -2.71748131e+02 -8.91704837e+01
  -2.67597261e+01]]

Intercept: 
 [-22.40828939]

R-squared:
0.9177575870306436


The model where the outcome Property Crime is predicted by the features population, robbery, murder and burglary explains 91.8% of the variance in Property Crime.Looking at the coefficients, Property Crime have the highest per-unit increase when population is more (3.52427802e-02).

In [80]:
# Write out the model formula.
# Your dependent variable on the right, independent variables on the left
# Use a ~ to represent an '=' from the functional form
linear_formula = 'Property_crime ~ Population+population_square+Robbery_new+Burglary_new+murder_new'

# Fit the model to our data using the formula.
lm = smf.ols(formula=linear_formula, data=df_temp).fit()

In [81]:
lm.params

Intercept            -22.408
Population             0.035
population_square     -0.000
Robbery_new         -271.748
Burglary_new         -89.170
murder_new           -26.760
dtype: float64

In [82]:
lm.pvalues

Intercept           0.989
Population          0.000
population_square   0.000
Robbery_new         0.348
Burglary_new        0.955
murder_new          0.869
dtype: float64

In [83]:
lm.rsquared

0.9177575870306436

From the p-values, we see the p-value for Robbery_new, Burglary_new, murder_new  is greater than .05, suggesting that there is probably no effect for those feature – and that dropping it would not adversely affect the  R2 .

In [99]:
data = df_temp[['Population']+['population_square']+['Robbery_new']+['Burglary_new']+['murder_new']]
target = df_temp['Property_crime']

In [91]:
# Test your model with different holdout groups.
# Use train_test_split to create the necessary training and test groups
X_train, X_test, y_train, y_test = train_test_split(data, target, test_size=0.2, random_state=20)
print('With 20% Holdout: ' + str(regr.fit(X_train, y_train).score(X_test, y_test)))

With 20% Holdout: 0.8353125910579849


In [96]:
X_test.shape

(93, 5)

In [92]:
from sklearn.model_selection import cross_val_score
cross_val_score(regr, X_train, y_train, cv=10)

array([0.87163678, 0.90755009, 0.76316451, 0.64479538, 0.79391367,
       0.49909281, 0.69982136, 0.33854673, 0.84498786, 0.77214954])

Lets drop those features and refit the model

In [84]:
# Write out the model formula.
# Your dependent variable on the right, independent variables on the left
# Use a ~ to represent an '=' from the functional form
linear_formula = 'Property_crime ~ Population+population_square'

# Fit the model to our data using the formula.
lm = smf.ols(formula=linear_formula, data=df_temp).fit()

In [85]:
lm.params

Intercept           -362.998
Population             0.035
population_square     -0.000
dtype: float64

In [86]:
lm.pvalues

Intercept           0.000
Population          0.000
population_square   0.000
dtype: float64

In [87]:
lm.rsquared

0.9175650939572668

In [98]:
data1 = df_temp[['Population']+['population_square']]
target1 = df_temp['Property_crime']

In [100]:
# Test your model with different holdout groups.
# Use train_test_split to create the necessary training and test groups
X_train, X_test, y_train, y_test = train_test_split(data1, target1, test_size=0.2, random_state=20)
print('With 20% Holdout: ' + str(regr.fit(X_train, y_train).score(X_test, y_test)))

With 20% Holdout: 0.8352116151398433


In [101]:
from sklearn.model_selection import cross_val_score
cross_val_score(regr, X_train, y_train, cv=10)

array([0.87082687, 0.91004664, 0.75981651, 0.64537477, 0.80687437,
       0.49627469, 0.70205014, 0.53295924, 0.84387879, 0.7713706 ])